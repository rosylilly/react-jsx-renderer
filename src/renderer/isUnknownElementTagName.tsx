export type UnknownHTMLElementTagNameFunction = (tagName: string) => boolean;

const unknownHTMLElementCache: Record<string, boolean> = {};
export const isUnknownHTMLElementTagName: UnknownHTMLElementTagNameFunction = (tagName) => {
  const cache = unknownHTMLElementCache[tagName];
  if (cache !== undefined) return cache;

  unknownHTMLElementCache[tagName] = global.document.createElement(tagName) instanceof global.window.HTMLUnknownElement;
  return !!unknownHTMLElementCache[tagName];
};
