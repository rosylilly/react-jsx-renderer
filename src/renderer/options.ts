import { Options } from '../types';
import { JSXElementFilter, JSXFragmentFilter, JSXTextFilter } from './filter';
import { UnknownHTMLElementTagNameFunction } from './isUnknownElementTagName';

export interface RenderingOptions extends Options {
  /**
   * List of filters to be applied to elements.
   */
  elementFilters?: JSXElementFilter[];

  /**
   * List of filters to be applied to fragments.
   */
  fragmentFilters?: JSXFragmentFilter[];

  /**
   * List of filters to be applied to text nodes.
   */
  textFilters?: JSXTextFilter[];

  /**
   * When this option is enabled, non-existent HTML elements will not be rendered.
   */
  disableUnknownHTMLElement?: boolean;

  /**
   * Function to determine Unknown HTML Element
   */
  isUnknownHTMLElementTagName?: UnknownHTMLElementTagNameFunction;
}
