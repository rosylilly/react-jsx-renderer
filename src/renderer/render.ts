import { ESTree } from 'meriyah';
import { createElement, Fragment, ReactNode } from 'react';
import { AnyFunction } from '..';
import { JSXElement, JSXFragment, JSXNode, JSXText } from '../types';
import { isUnknownHTMLElementTagName } from './isUnknownElementTagName';
import { RenderingOptions } from './options';

const fileName = 'jsx';

export const renderJSX = (node: JSXNode | JSXNode[], options: RenderingOptions): ReactNode | ReactNode[] => {
  if (node === null) return node;
  if (node === undefined) return node;
  if (Array.isArray(node)) return node.map((n) => renderJSX(n, options));

  switch (typeof node) {
    case 'boolean':
      return node;
    case 'string':
    case 'number':
      return renderJSXText(node, options);
    default:
      return renderJSXNode(node, options);
  }
};

const renderJSXText = (text: JSXText, options: RenderingOptions): ReactNode => {
  return applyFilter(options.textFilters || [], text);
};

const renderJSXNode = (node: JSXElement | JSXFragment, options: RenderingOptions): ReactNode => {
  switch (node.type) {
    case 'element':
      return renderJSXElement(node, options);
    case 'fragment':
      return renderJSXFragment(node, options);
  }
};

const renderJSXElement = (element: JSXElement, options: RenderingOptions): ReactNode => {
  const filtered = applyFilter(options.elementFilters || [], element);
  if (!filtered) return undefined;

  if (options.disableUnknownHTMLElement && typeof filtered.component === 'string') {
    const { component } = filtered;
    const checker = options.isUnknownHTMLElementTagName || isUnknownHTMLElementTagName;
    if (checker(component)) return undefined;
  }

  const component = filtered.component as AnyFunction;
  if (typeof component === 'function') {
    const props = { ...filtered.props };
    let children = component(props);
    if (!Array.isArray(children)) children = [children];
    return createElement(Fragment, { ...renderSourcePosition(element.loc, options) }, ...children.map((child) => renderJSX(child, options)));
  }

  return createElement(
    filtered.component,
    {
      ...filtered.props,
      ...renderSourcePosition(element.loc, options),
    },
    ...filtered.children.map((child) => renderJSX(child, options)),
  );
};

const renderJSXFragment = (fragment: JSXFragment, options: RenderingOptions): ReactNode => {
  const filtered = applyFilter(options.fragmentFilters || [], fragment);

  if (filtered) {
    return createElement(
      Fragment,
      {
        ...filtered.props,
        ...renderSourcePosition(fragment.loc, options),
      },
      ...filtered.children.map((child) => renderJSX(child, options)),
    );
  } else {
    return undefined;
  }
};

const applyFilter = <T extends JSXNode>(filters: ((target: T) => T | undefined)[], node: T): T | undefined => {
  return filters.reduce<T | undefined>((prev, filter) => (prev ? filter(prev) : undefined), node);
};

type SourcePosition = {
  __source: {
    fileName: string;
    lineNumber?: number;
    columnNumber?: number;
  };
};

const renderSourcePosition = (loc: ESTree.Position | undefined, _options: RenderingOptions): SourcePosition | Record<string, never> => {
  return loc ? { __source: { fileName, lineNumber: loc.line, columnNumber: loc.column } } : { __source: { fileName } };
};
