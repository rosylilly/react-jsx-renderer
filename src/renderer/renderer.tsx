import { ESTree } from 'meriyah';
import React, { createContext, FC, memo, Ref, useContext, useEffect, useMemo, VFC } from 'react';
import { evaluateJSX, EvaluateOptions, parse, ParseOptions } from '../evaluate';
import { JSXNode } from '../types';
import { RenderingOptions } from './options';
import { renderJSX } from './render';

export type JSXFallbackComponent = VFC<{ error: Error } & JSXRendererProps>;

export interface JSXRendererProps extends ParseOptions, EvaluateOptions, RenderingOptions {
  /**
   * JSX code
   */
  code?: string;

  /**
   * The component that will be displayed instead when an error occurs.
   */
  fallbackComponent?: JSXFallbackComponent;

  /**
   * If you want to receive the parsed result, set a Ref object to this option.
   */
  refNodes?: Ref<JSXNode[]>;
}

const JSXRendererContext = createContext<JSXRendererProps>({});
export const JSXRendererOptionsProvider: FC<JSXRendererProps> = ({ children, ...props }) => {
  return <JSXRendererContext.Provider value={props}>{children}</JSXRendererContext.Provider>;
};

const DefaultJSXFallbackComponent: JSXFallbackComponent = (props) => {
  const { error, debug } = props;
  debug && console.error(error);

  return <>{error.message}</>;
};
DefaultJSXFallbackComponent.displayName = 'DefaultJSXFallbackComponent';

const Renderer: VFC<JSXRendererProps> = (props: JSXRendererProps) => {
  const contextOptions = useContext(JSXRendererContext);
  const { code, fallbackComponent, refNodes, ...options } = Object.assign({}, contextOptions, props);
  const { meriyah, debug } = options;
  const Fallback = fallbackComponent ? fallbackComponent : DefaultJSXFallbackComponent;

  let nodes: JSXNode[] = [];

  debug && console.group('JSXRenderer');

  const [program, error] = useMemo(() => {
    let program: ESTree.Program | undefined = undefined;
    let error: Error | undefined = undefined;

    try {
      program = parse(code || '', { meriyah, debug, forceExpression: true });
    } catch (e) {
      error = e;
    }
    return [program, error];
  }, [code, meriyah, debug]);

  useEffect(() => {
    if (typeof refNodes === 'function') refNodes(nodes);
  }, [refNodes, program]);

  try {
    if (program) nodes = evaluateJSX(program, options);

    return (
      <>
        {error && <Fallback error={error} />}
        {nodes.map((node) => renderJSX(node, options))}
      </>
    );
  } catch (err) {
    const error = err;

    return <Fallback {...props} error={error} />;
  } finally {
    debug && console.groupEnd();
  }
};
Renderer.displayName = 'JSXRenderer';

export const JSXRenderer = memo(Renderer);
