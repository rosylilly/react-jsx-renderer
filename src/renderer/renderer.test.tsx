import React, { FC, ReactElement } from 'react';
import renderer from 'react-test-renderer';
import { JSXRenderer } from '.';
import { JSXElementFilter, JSXFragmentFilter, JSXTextFilter } from './filter';
import { JSXFallbackComponent, JSXRendererOptionsProvider } from './renderer';
import { JSDOM } from 'jsdom';

describe('JSXRenderer', () => {
  if (global.document === undefined) {
    // Node environment only
    const { window } = new JSDOM();
    global.window = window as any;
    global.document = window.document;
  }

  const mockConsoleError = jest.spyOn(global.console, 'error').mockImplementation();
  const mockConsoleGroup = jest.spyOn(global.console, 'group').mockImplementation();
  const mockConsoleGroupEnd = jest.spyOn(global.console, 'groupEnd').mockImplementation();
  const mockConsoleTime = jest.spyOn(global.console, 'time').mockImplementation();
  const mockConsoleTimeEnd = jest.spyOn(global.console, 'timeEnd').mockImplementation();

  afterEach(() => {
    mockConsoleError.mockClear();
    mockConsoleGroup.mockClear();
    mockConsoleGroupEnd.mockClear();
    mockConsoleTime.mockClear();
    mockConsoleTimeEnd.mockClear();
  });

  it('should render JSX', () => {
    const tree = renderer
      .create(
        <JSXRenderer
          code={`
      <h1>Hello, JSX World</h1>
      <p>This is paragraph.</p>
      <ul className="Hello">
        <li>first</li>
        <li>second</li>
      </ul>
      <>Fragment</>
    `}
          keyPrefix="html"
        />,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should catch error of syntax', () => {
    const tree = renderer.create(<JSXRenderer disableKeyGeneration code={`<p ...`} />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should catch error with number object', () => {
    const tree = renderer
      .create(
        <JSXRenderer
          disableKeyGeneration
          code={`<p>{raise()}</p><p />`}
          binding={{
            raise: () => {
              throw 1;
            },
          }}
        />,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should catch error with object', () => {
    const tree = renderer
      .create(
        <JSXRenderer
          disableKeyGeneration
          code={`<p>{raise()}</p>`}
          binding={{
            raise: () => {
              throw { foo: 1 };
            },
          }}
        />,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  const elementFilters: JSXElementFilter[] = [(e) => (e.component === 'link' ? undefined : e), (e) => (e.component === 'script' ? undefined : e)];
  const fragmentFilters: JSXFragmentFilter[] = [
    (fragment) => {
      return fragment.children[0] === 'delete' ? undefined : fragment;
    },
  ];
  const textFilters: JSXTextFilter[] = [(text) => (text === 'ban' ? undefined : text)];
  const fallback: JSXFallbackComponent = ({ error }) => <div>{error.message} is raised</div>;
  const test = (code: string, expected: ReactElement) => {
    it(`should render ${code}`, () => {
      const actual = renderer
        .create(
          <JSXRenderer
            code={code}
            binding={{
              exception() {
                throw new Error('Hello');
              },
            }}
            elementFilters={elementFilters}
            fragmentFilters={fragmentFilters}
            textFilters={textFilters}
            fallbackComponent={fallback}
            disableSearchCompontsByBinding
            disableUnknownHTMLElement
          />,
        )
        .toJSON();
      expect(actual).toStrictEqual(renderer.create(expected).toJSON());
    });
  };

  test('', <></>);
  test('<p>Hoge</p>', <p>Hoge</p>);
  test('<p>{true}</p>', <p>{true}</p>);
  test('<p>{null}</p>', <p>{null}</p>);
  test('<p>{undefined}</p>', <p>{undefined}</p>);
  test(
    '<p><>frag</></p>',
    <p>
      <>frag</>
    </p>,
  );
  test('<p><>delete</></p>', <p></p>);
  test('<p><script src="./index.js" /></p>', <p></p>);
  test('<p><link href="./index" /></p>', <p></p>);
  test('<p><unknown>this is removed</unknown>element</p>', <p>element</p>);
  test(
    '<p><b>ban</b>word</p>',
    <p>
      <b />
      word
    </p>,
  );
  test('<p>{exception()}</p>', <div>{'Hello'} is raised</div>);
  test(
    '<p>{["a", "b", "c"]}</p>',
    <p>
      {'a'}
      {'b'}
      {'c'}
    </p>,
  );

  it('should call console.error on debug mode', () => {
    const tree = renderer.create(<JSXRenderer code="<span>{foo()}</span>" debug />).toJSON();
    expect(tree).toMatchSnapshot();

    expect(mockConsoleError).toHaveBeenCalledTimes(1);
    expect(mockConsoleGroup).toHaveBeenCalledTimes(1);
    expect(mockConsoleGroupEnd).toHaveBeenCalledTimes(1);
    expect(mockConsoleTime).toHaveBeenCalledTimes(2);
    expect(mockConsoleTimeEnd).toHaveBeenCalledTimes(2);
  });

  it('should call console.group on debug mode', () => {
    const tree = renderer.create(<JSXRenderer code="<p>{1}</p>" debug />).toJSON();
    expect(tree).toMatchSnapshot();

    expect(mockConsoleGroup).toHaveBeenCalledTimes(1);
    expect(mockConsoleGroupEnd).toHaveBeenCalledTimes(1);
    expect(mockConsoleTime).toHaveBeenCalledTimes(2);
    expect(mockConsoleTimeEnd).toHaveBeenCalledTimes(2);
  });

  it('should call refNodes with node list', () => {
    const refNodes = jest.fn(() => {});
    const { act } = renderer;
    act(() => {
      renderer.create(<JSXRenderer code="{1} is 1" refNodes={refNodes} />).toJSON();
    });

    expect(refNodes).toBeCalledWith([1, ' is 1']);
  });

  it('should call Fallback Component on parse error', () => {
    const fallbackComponent = jest.fn(({ error }) => <>{error.message}</>);
    const tree = renderer.create(<JSXRenderer code="{notFound()}" fallbackComponent={fallbackComponent} />).toJSON();
    expect(tree).toMatchSnapshot();

    expect(fallbackComponent).toHaveBeenCalledWith(
      expect.objectContaining({
        code: '{notFound()}',
        fallbackComponent,
        error: expect.any(Error),
      }),
      {},
    );
  });

  it('should override with context', () => {
    const tree = renderer
      .create(
        <JSXRendererOptionsProvider binding={{ hello: 'by context' }}>
          <JSXRenderer code="{hello}" />
        </JSXRendererOptionsProvider>,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should render with custom component', () => {
    const code = `
      <Hello>World</Hello>
    `;
    const Hello: FC = ({ children }) => {
      return <>Hello, {children}</>;
    };
    const tree = renderer
      .create(
        <JSXRendererOptionsProvider binding={{ hello: 'by context' }} components={{ Hello }}>
          <JSXRenderer code={code} />
        </JSXRendererOptionsProvider>,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should render with component name', () => {
    const code = `
      export const B = function(){ return 'other function' };
      export const Main = () => <>Hello, World: {hello} / {B()}</>;
    `;
    const tree = renderer
      .create(
        <JSXRendererOptionsProvider binding={{ hello: 'by context' }}>
          <JSXRenderer code={code} component="Main" />
        </JSXRendererOptionsProvider>,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should render empty on non function component', () => {
    const code = `
      export const B = 'Hello'
    `;
    const tree = renderer
      .create(
        <JSXRendererOptionsProvider binding={{ hello: 'by context' }}>
          <JSXRenderer code={code} component="B" />
        </JSXRendererOptionsProvider>,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
