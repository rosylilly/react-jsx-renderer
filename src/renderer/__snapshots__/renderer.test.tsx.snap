// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`JSXRenderer should call Fallback Component on parse error 1`] = `"notFound is not a function"`;

exports[`JSXRenderer should call console.error on debug mode 1`] = `"[1:9] foo is not a function"`;

exports[`JSXRenderer should call console.group on debug mode 1`] = `
<p>
  1
</p>
`;

exports[`JSXRenderer should catch error of syntax 1`] = `"[1:11]: Unexpected token"`;

exports[`JSXRenderer should catch error with number object 1`] = `"1"`;

exports[`JSXRenderer should catch error with object 1`] = `"[object Object]"`;

exports[`JSXRenderer should override with context 1`] = `"by context"`;

exports[`JSXRenderer should render JSX 1`] = `
Array [
  "
      ",
  <h1>
    Hello, JSX World
  </h1>,
  "
      ",
  <p>
    This is paragraph.
  </p>,
  "
      ",
  <ul
    className="Hello"
  >
    
        
    <li>
      first
    </li>
    
        
    <li>
      second
    </li>
    
      
  </ul>,
  "
      ",
  "Fragment",
]
`;
